kufar/books/main.go

```diff
@@ -103,19 +120,79 @@ func (d Database) Close() error {
 	return d.db.Close()
 }
 
+//SignUpUser ...
+func (d Database) SignUpUser(w http.ResponseWriter, r *http.Request) {
+	credReq := CredentialsRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&credReq); err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+"Error unmarshalling json", http.StatusInternalServerError)
+		return
+	}
+
+	const cost = 12
+	pword, err := bcrypt.GenerateFromPassword([]byte(credReq.Password), cost)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	cred := Credentials{Username: credReq.Username}
+	if _, err = d.db.Exec("insert into credentials(username, password) values($1, $2)", cred.Username, string(pword)); err != nil {
+		http.Error(w, http.StatusText(http.StatusInternalServerError)+": could not log in user. Try different uuid", http.StatusInternalServerError)
+		return
+	}
+
+	w.Write([]byte(http.StatusText(http.StatusOK)))
+}
+
+var store = sessions.NewCookieStore([]byte(getenv("SESSION_KEY", "session_key")))
```

Comment:
Its better to move it up.
Basically, package file consists of (from top to bottom):
1. Imports
2. Consts
3. Global Vars
4. Types
5. Functions

Reply:
This will be very helpful. Cheers ;)